storehaus-cassandra/src/main/scala/com/twitter/storehaus/cassandra/cql/AbstractCQLCassandraSimpleStore.scala

```diff
@@ -0,0 +1,126 @@
+import com.datastax.driver.core.{BatchStatement, ConsistencyLevel, ResultSet, Row, SimpleStatement, Statement}
+import com.twitter.concurrent.Spool
+import com.twitter.storehaus.{IterableStore, QueryableStore, ReadableStore, Store}
+import com.twitter.util.{Duration, Future, FuturePool, Promise, Throw, Return}
+import java.util.concurrent.{Executors, TimeUnit}
+import org.slf4j.{ Logger, LoggerFactory }
+import com.websudos.phantom.CassandraPrimitive
+import com.twitter.util.Time
+import com.datastax.driver.core.VersionNumber
+import com.twitter.util.Try
+
+abstract class AbstractCQLCassandraSimpleStore[K : CassandraPrimitive, V] (
+    override val poolSize: Int, 
+    override val columnFamily: CQLCassandraConfiguration.StoreColumnFamily,
+    keyColumnName: String = CQLCassandraConfiguration.DEFAULT_KEY_COLUMN_NAME,
+    consistency: ConsistencyLevel = CQLCassandraConfiguration.DEFAULT_CONSISTENCY_LEVEL,
+    batchType: BatchStatement.Type = CQLCassandraConfiguration.DEFAULT_BATCH_STATEMENT_TYPE,
+    ttl: Option[Duration] = CQLCassandraConfiguration.DEFAULT_TTL_DURATION) 
+  extends AbstractCQLCassandraStore[K, V](poolSize, columnFamily)
+  with Store[K, V] {
+  
+  @transient private val log = LoggerFactory.getLogger(classOf[AbstractCQLCassandraSimpleStore[K, V]])
+
+  log.debug(s"""Creating new AbstractCQLCassandraSimpleStore on ${columnFamily.session.cluster.hosts} with consistency=${consistency.name} and 
+    load-balancing=${columnFamily.session.getCluster.getConfiguration.getPolicies.getLoadBalancingPolicy}""")
+
+  val keySerializer = implicitly[CassandraPrimitive[K]]
+  
+  protected lazy val cassandraVersion: Option[VersionNumber] =
+    // assumes that all hosts are the same version and there is at least one host available
+    Try(columnFamily.session.getCluster.getMetadata.getAllHosts.iterator.next.getCassandraVersion).toOption
+
+  
+  protected def deleteColumns: Option[String]
+  
+  protected def createPutQuery[K1 <: K](kv: (K1, V)): Insert
+
+  override def getKeyValueFromRow(row: Row): (K, V)
+  
+  override def multiPut[K1 <: K](kvs: Map[K1, Option[V]]): Map[K1, Future[Unit]] = {
+    if(kvs.size > 0) {
+      val result = futurePool[Unit] {
+    	val mutator = new BatchStatement(batchType)
+    	kvs.foreach {
+    	  case (key, Some(value)) => { 
+    	    val builder = createPutQuery((key, value))
+    	    ttl match {
+    	      case Some(duration) => builder.using(QueryBuilder.ttl(duration.inSeconds))
+    	      case _ =>
+    	    }
+    	    mutator.add(builder)
+    	  }
+    	  case (key, None) => mutator.add(deleteColumns match {
+    	    case None => QueryBuilder.delete().from(columnFamily.getName).where(QueryBuilder.eq(keyColumnName, key))
+    	    case Some(cols) => QueryBuilder.delete(cols).from(columnFamily.getName).where(QueryBuilder.eq(keyColumnName, key))
+    	  })
+    	}
+    	mutator.setConsistencyLevel(consistency)
+    	columnFamily.session.getSession.execute(mutator)
```

Comment:
is this thread-safe? Should we use the AsyncMutex to ensure only one query is running?